### ISSUE TITLE
snakefmt complains on its own formatting on multiple if statements

### ISSUE THREAD

**Issue by weber8thomas** [snakemake/snakefmt#157]
I'm experiencing an issue on a multiple conditions if statement using snakefmt

### 3 if statements on a single line

![Screenshot 2022-11-08 at 10 37 41](https://user-images.githubusercontent.com/32496011/200530171-ae0089f0-7b88-43f7-9a4a-023539d4b5e2.png)

### snakefmt works fine
![Screenshot 2022-11-08 at 10 38 23](https://user-images.githubusercontent.com/32496011/200530176-b44a3926-d48f-4eae-9122-170a24c39ae8.png)

### New formatted code looks like this
![Screenshot 2022-11-08 at 10 38 54](https://user-images.githubusercontent.com/32496011/200530180-1b28500d-5ac4-4b74-b17d-21f1587d6225.png)

### snakefmt now complains on its own formatting

![Screenshot 2022-11-08 at 10 39 14](https://user-images.githubusercontent.com/32496011/200530185-0b2e7034-d1bb-4186-ac70-babb9f4d77bf.png)

**Comment by mbhall88** [comment]
Could you paste the original code in a code fence here so I can reproduce and debug this?

**Comment by corneliusroemer** [comment]
I can reproduce, this is how it cycles in 0.7.0:

Start:
```
if config.get("s3_dst"):

    include: "workflow/snakemake_rules/upload.smk"
    include: "workflow/snakemake_rules/trigger.smk"
    # Only include rules for trigger if uploading files since the trigger
    # rules depend on the outputs from upload.
    include: "workflow/snakemake_rules/trigger.smk"
```

After formatting:

```
if config.get("s3_dst"):

    include: "workflow/snakemake_rules/upload.smk"
    include: "workflow/snakemake_rules/trigger.smk"


# Only include rules for trigger if uploading files since the trigger
# rules depend on the outputs from upload.
include: "workflow/snakemake_rules/trigger.smk"
```

But this is still not ok per Snakefmt, if I save again:

```
if config.get("s3_dst"):

    include: "workflow/snakemake_rules/upload.smk"
    include: "workflow/snakemake_rules/trigger.smk"
# Only include rules for trigger if uploading files since the trigger
# rules depend on the outputs from upload.
include: "workflow/snakemake_rules/trigger.smk"
```

**Comment by weber8thomas** [comment]
> Could you paste the original code in a code fence here so I can reproduce and debug this?

@mbhall88 Sure, wanted to avoid some formatting changes by making a screenshot

Here is the code:

Before formatting:

```python
if config["window"] in [50000, 100000, 200000] and config["reference"] == "hg38" and config["normalized_counts"] == True:

    rule merge_blacklist_bins:
        input:
            norm="workflow/data/normalization/HGSVC.{window}.txt",
            whitelist="workflow/data/normalization/inversion-whitelist.tsv",
        output:
            merged="{folder}/{sample}/normalizations/HGSVC.{window}.merged.tsv",
        log:
            "{folder}/log/normalizations/{sample}/HGSVC.{window}.merged.tsv",
        conda:
            "../envs/mc_base.yaml"
        shell:
            """
            workflow/scripts/normalization/merge-blacklist.py --merge_distance 500000 {input.norm} --whitelist {input.whitelist} --min_whitelist_interval_size 100000 > {output.merged} 2>> {log}
            """
```

After formatting

```python
if (
    config["window"] in [50000, 100000, 200000]
    and config["reference"] == "hg38"
    and config["normalized_counts"] == True
):

    rule merge_blacklist_bins:
        input:
            norm="workflow/data/normalization/HGSVC.{window}.txt",
            whitelist="workflow/data/normalization/inversion-whitelist.tsv",
        output:
            merged="{folder}/{sample}/normalizations/HGSVC.{window}.merged.tsv",
        log:
            "{folder}/log/normalizations/{sample}/HGSVC.{window}.merged.tsv",
        conda:
            "../envs/mc_base.yaml"
        shell:
            """
            workflow/scripts/normalization/merge-blacklist.py --merge_distance 500000 {input.norm} --whitelist {input.whitelist} --min_whitelist_interval_size 100000 > {output.merged} 2>> {log}
            """
```

@corneliusroemer I also experienced the same issue regarding comments block (""") or comments lines (#)

snakefmt version used: 0.6.1

**Comment by mbhall88** [comment]
As mentioned in #161, we (Brice and I) are just discussing who can address what when. I am on holidays next week and he has is PhD defence, so it will likely be after next week we get around to these sorry. #159 seems to be the first priority, but some of these other issues might be wrapped up in the solution for that bug.

Thank you very much for the bug reports!

**Comment by weber8thomas** [comment]
> As mentioned in #161, we (Brice and I) are just discussing who can address what when. I am on holidays next week and he has his PhD defence, so it will likely be after next week we get around to these sorry. #159 seems to be the first priority, but some of these other issues might be wrapped up in the solution for that bug.
> 
> Thank you very much for the bug reports!

No urgence on my side, just wanted to report it 

Good luck to @bricoletc for his PhD defence and enjoy you holidays @mbhall88 !

**Comment by johanneskoester** [comment]
I think #159 is quite critical indeed. I have no experience with the snakefmt codebase myself, but if you give me a pointer where to start, I can try to help here.

**Comment by corneliusroemer** [comment]
I had a go at trying to debug this stuff myself over the weekend but gave up after half a dozen hours. I managed to get a simple new test case to pass, but then when I used two comment lines rather than one things failed again.

The tests should be a useful guide along the way helping with a reimplementation.

Here are test cases for #159:

```py
    def test_if_statement_with_snakecode_comment_snakecode_inside(self):
        """https://github.com/snakemake/snakefmt/issues/159"""
        snakecode = (
            "if True:\n\n"
            f"{TAB * 1}ruleorder: a > b\n"
            "\n"
            f"# comment\n"
            f"{TAB * 1}ruleorder: c > d\n"
        )
        formatter = setup_formatter(snakecode)
        assert formatter.get_formatted() == snakecode

    def test_if_statement_with_snakecode_2comments_snakecode_inside(self):
        """https://github.com/snakemake/snakefmt/issues/159"""
        snakecode = (
            "if True:\n\n"
            f"{TAB * 1}ruleorder: a > b\n"
            "\n"
            f"# comment\n"
            f"# comment\n"
            f"{TAB * 1}ruleorder: c > d\n"
        )
        formatter = setup_formatter(snakecode)
        assert formatter.get_formatted() == snakecode
```

**Comment by mbhall88** [comment]
> Perhaps trying to implement an AST-based parser/formatter would be much easier for maintanibility.
This issue is not the place for such a discussion btw, not sure where would be best ;)

Agreed. This is what I wanted to do in the first place 😛 

But that would require a significant amount of work that I don't have and I doubt @bricoletc does either. Maybe @johanneskoester has someone in his team that could do this?

Once I get #163 and #164 fixed, I will look at this. It is quite possible #164 could impact this issue

**Comment by bricoletc** [comment]
> I had a go at trying to debug this stuff myself over the weekend but gave up after half a dozen hours. I managed to get a simple new test case to pass, but then when I used two comment lines rather than one things failed again.

A valiant effort @corneliusroemer !

I'd be up to having a go at a CST/AST-based rewrite, if others, like @corneliusroemer /someone from @johanneskoester 's team can join in.

**Comment by johanneskoester** [comment]
The rewrite plans sound good, but also as if that will need a lot of work. So, if we can fix this before, I think that would be good.
I might be able to dedicate some person hours to something like that later in 2023, but as there are many other construction sites, it would be highly appreciated if you would want to start before.

**Comment by mbhall88** [comment]
The examples in this issue are resolved in v0.8.0
